rust/connlib/tunnel/src/role_state.rs

```diff
@@ -0,0 +1,146 @@
+use crate::Event;
+use connlib_shared::messages::{ClientId, GatewayId};
+use futures::channel::mpsc::Receiver;
+use futures_bounded::{PushError, StreamMap};
+use std::collections::HashMap;
+use std::fmt;
+use std::task::{ready, Context, Poll};
+use std::time::Duration;
+use webrtc::ice_transport::ice_candidate::RTCIceCandidateInit;
+
+/// Dedicated trait for abstracting over the different ICE states.
+///
+/// By design, this trait does not allow any operations apart from advancing via [`RoleState::poll_next_event`].
+/// The state should only be modified when the concrete type is known, e.g. [`ClientState`] or [`GatewayState`].
+pub trait RoleState: Default + Send + 'static {
+    type Id: fmt::Debug;
+
+    fn poll_next_event(&mut self, cx: &mut Context<'_>) -> Poll<Event<Self::Id>>;
+}
+
+/// For how long we will attempt to gather ICE candidates before aborting.
+///
+/// Chosen arbitrarily.
+const ICE_GATHERING_TIMEOUT_SECONDS: u64 = 5 * 60;
```

Comment:
We should coordinate this with `webrtc`

Reply:
Are you happy to just leave it for now? I think 5 minutes is pretty insane and for a good UX, this should be a lot less. I think any successful connection will be done before that and only if something goes wrong, we make sure that this is cleaned up.

Reply:
Oh yeah, I'd just leave a todo here or an explanation that there's an independent timeout in webrtc that will always timeout before this.